students/ptfbanks/Lesson05/pf_mailroom-3.py

```diff
@@ -0,0 +1,123 @@
+   "It will be put to very good use.\n\n"
+   "\t\t Sincereley,\n"
+   "\t\t   -The Team")
+
+#-------------------------------Functions----------------------#
+
+def hist_rpt():   #Prepare list of donors, sorted by sum donation
+    print("|   DonorName    | TotaL Given | Num Gifts   | Avrage Gift |")
+    print("|----------------|-------------|-------------|-------------|")
+    d_srt = {name: sum(d_dict[name])for name in d_dict}     #Compression for:
+#    d_srt = {}
+#    for name in d_dict:
+#        tot = sum(d_dict[name])
+#        d_srt[name]=tot
+    ordered = sorted(d_srt, key=d_srt.__getitem__, reverse=True)
+    for name in ordered:
+        t_val=d_srt[name]
+        n_o_d=len(d_dict[name])
+        s_avg= t_val/n_o_d
+        print("|{:^16}|{:>13,.02f}|{:^13}|{:>13,.02f}|".format(name, t_val, n_o_d, s_avg))
+        print("|----------------|-------------|-------------|-------------|")        
+    
+def menu_select(choice):  #select and apply menu  (Excepton check incluced)
+    try:
+        while True:
+            pick=input(choice)
+            if pick =='4':
+                print('So long.....')
+                break
+            menu[pick]()
+    except KeyError:
+        print('command not understood.  Plese try again.\n')
+        menu_select(choice)
+
+def send_where (name, send, direct):
+    amt = sum(d_dict[name])  #Note: Need to reslove decimal formatting for letter
+    if send.lower() == 'y':
+        destination=os.path.join(direct,"{}.txt".format(name))
+        with open(destination, 'w') as f:
+            f.write(letter.format(name, amt))
+    else:   #screen display of all T/U letters
+        print(letter.format(name, amt))
+        
+def thank_you():  # Prepare single or collective set of thank_ you letters 
+    direct = ""
+    send = input('Do you want Thank You letters stored as .txt files in a directory? (y/n):')
+    if send.lower() == 'y':
+        print('OK, a .txt file will be generated. See directory for text files generated.\n')
+#       direct = input('Type directory name (retrun blank for current directory):')
+#  The above line is retained for future design to expand file sending optoins.
+    else:
+        print('OK, text is be printed on screen below for copy and pasting into an external file.')
+    whom = input('For whom do you want to prepare a Thank You letter?  Enter a name or "All":')
+    if whom.lower() == 'all':   #Prepare text of T/U letters for all donors showing summary amounts
+        for name in d_dict:
+            send_where (name, send, direct)
+    else:                                       #(Excepton check incluced)
+        try:
+            #print("test", whom, " - ", send, "---", direct, "+++") 
+            send_where(whom, send, direct)
+        except KeyError:
+            print("\n There is no record of donations by: ", whom, ".  Refresh donor report and try again.")
+            menu_select(choice)
```

Comment:
don't call the main loop manually; you're in a while True so it will loop naturally